### PR TITLE
Docs build overhaul

### PR DESCRIPTION
**Pull request by smackesey**
### Summary & Motivation

This PR updates and cleans up our docs code and README with a particular focus on Sphinx and the Screenshot management code. It attempts to remove superfluous or outdated config/extensions, ease the workflow for iterating on API docs, update to newer versions of sphinx and deps, and thoroughly explain how everything fits together in the README.

There are a lot of changes listed, and it's possible I didn't understand everything, so anything here is of course open to reversion.

- README
    - Reorganized and added a lot of explanatory content.
- API docs (sphinx) build environment:
    - Removed `py38` factor from `sphinx` tox environment. This means sphinx will attempt to use whatever Python version you're using (as of now it works on 3.8, 3.9, and 3.10). This was mainly for consistency with other tox configurations throughout the dagster codebase.
    - Bumped sphinx version from 3.5.2 to 5.0.2
    - Bumped sphinx-click version from 2.5.0 to 4.3.0
    - Removed nbsphinx package/extension. We do not use it.
    - Removed recommonmark extension. This is used to parse `.md` files alongside `.rst` files, but we don't actually have any `.md` files in our API docs.
    - Delete `docs-requirements.txt` and no longer install all target modules (i.e. dagster libs) into the environment with `pip`. Instead, place them on `sys.path` in `conf.py`. This is sphinx recommended way of making them available to autodoc. It also allows us to build the sphinx environment faster and more predictably (with a single `pip install` command instead of the previously required two due to use of `--no-deps`).
    - Delete `docs-build-requirements.txt`, fold its content into the `sphinx` tox env dependencies, and edit the list. In particular I removed many one-off dependencies where it was unclear why they were there, and replaced them with editable installs of the dagster package they service. There are two reasons it is preferable to keep the list in the tox env deps instead of a requirements file: (1) it allows tox to track changes to the dependencies and only rebuild the sphinx env when needed; (2) it's simpler.
- Sphinx config:
    - Remove `nbsphinx`, `recommonmark`, `intersphinx`, `ifconfig` extensions.
    - Remove a large amount of boilerplate config related to HTML and latex output. Pretty sure this stuff was added by the config generator for an old version of sphinx, but we don't use any of it since we only build JSON.
    - Removed a bunch of unnecessary entries from `exclude`. The patterns here matched various JS items but `exclude_patterns` is relative to the API doc source root, which does not have any JS stuff.
    - Removed HTML templates in `_templates`, since we don't use them.
- Next
    - Some cosmetic changes to `yarn snapshot`, `yarn format` and associated script, now the main task is `yarn mdx-format` (to match `Makefile`).
- Makefile:
    - Rename `build` task to `apidoc-build` to better reflect that it is building the API docs rather than the doc site
    - Folded sphinx part of `make build` from `Makefile` into `testenv:sphinx` commands. There is no reason to keep `make build` outside the tox file since it needs to run within the environment defined in the toxfile.
    - Added `apidoc-watch-build`, which uses watchdog's CLI script `watchmedo` to watch the APIdoc rst files and config and rebuild on changes. Building is kinda slow, but you can use this together with `make next-watch-build` (you need two separate processes/terminals) to get live-reloading of APIdocs in the dev server.
    - Rename `dev` task to `next-watch-build` to better specify this task is specific to `next`.
    - Rename `docs_dev_install` to `next-dev-install` to better specify this task is specific to `next`.
    - Rename `snapshot` to `mdx-format` to better describe what the task does.
- Toxfile adjustments
    - Removed common deps from `testenv`. All environments have separate dep requirements.
    - Removed `pip install -r ...` from `commands` section for sphinx build. Placed in deps section under `[testenv:sphinx]`.
- Buildkite
    - ~~Changed the docs build step so that it checks not only that the API docs build, but also that the build output matches what's checked in to the repo (using `git diff --exit-code`). This prevents merging when you forgot to rebuild API docs after changes.~~
- Screenshots
    - Screenshot code has been refactored and placed into a Python package `dagit-screenshot`, which can be installed into the dev Python environment. This installs the executable `dagit-screenshot`, which is a `click` CLI with commands `capture` (formerly the script `capture-screenshot.py`) and `audit` (formerly `match_screenshots.py`) and `show` (new-- dumps contents of screenshot database). Packaging this code into a package provides a base for further development of our screenshot automation.
    - `dagit-screenshot/README.md` describes how to use it in detail.
    - Screenshot specs now specify `route` and `base_url` separately instead of a full URL-- `base_url` defaults to `http://localhost:3000`
    - `path` field of screenshot spec now called `id`. Output path is by default derived from the `id`, but can be overridden in the CLI. Useful for experimenting without overwriting existing images.
    - `screenshots.yaml` fully decomposed into a multi-file hierarchy rooted in `docs/screenshots` that mirrors the layout of `docs/content`. Someone had previously partially started this process, but now its complete.
- Misc
    - Moved `pack_json.py` into a "scripts" directory.

### How I Tested These Changes

BK

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #8960** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/8960" target="_blank">Graphite</a>  👈
    * **PR #8961** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/8961" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/8960?utm_source=stack-comment).

File: docs/sphinx/conf.py [hunk 1]
```diff
@@ -96,158 +146,16 @@
...
-        "Dagster",
-        "One line description of project.",
-        "Miscellaneous",
-    )
-]
-
-# -- Extension configuration -------------------------------------------------
-
-extensions = extensions + ["sphinx.ext.napoleon", "sphinx_click.ext"]
+# ????
```
Comment (yuhan) [inline docs/sphinx/conf.py:151]: rm?

File: .buildkite/dagster-buildkite/dagster_buildkite/steps/docs.py [hunk 2]
```diff
@@ -19,39 +19,38 @@ def build_docs_steps() -> List[BuildkiteStep]:
...
         .build(),
         # Make sure docs sphinx build runs.
-        CommandStepBuilder("docs sphinx json build")
+        CommandStepBuilder("docs sphinx build")
         .run(
             "pip install -U virtualenv",
             "cd docs",
-            "tox -vv -e py38-sphinx",
+            "tox -vv -e sphinx",
+            "git diff --exit-code",
```
Comment (yuhan) [inline .buildkite/dagster-buildkite/dagster_buildkite/steps/docs.py:42]: 🙏

**Review by yuhan — APPROVED** [review]
🫡 thank you for service. this is lord's work

**Pull request merged**